content/document.js

```diff
@@ -30,6 +30,85 @@ const {
 } = require("./utils");
 const Redirect = require("./redirect");
 
+function allDocumentPathsAsTree(root) {
+  const api = new fdir()
+    .withErrors()
+    .withRelativePaths()
+    .filter((filePath) => {
+      return (
+        filePath.endsWith(HTML_FILENAME) || filePath.endsWith(MARKDOWN_FILENAME)
+      );
+    })
+    .crawl(root);
+
+  const allPaths = [...api.sync()].map((p) => p.split(path.sep));
+  let tree = {};
+  for (const paths of allPaths) {
+    tree = addToTree(tree, paths);
+  }
+
+  return tree;
+}
+function addToTree(tree = {}, [current, ...rest]) {
+  if (rest.length === 0) {
+    tree[current] = null;
+    return tree;
+  }
+  tree[current] = addToTree(tree[current], rest);
+  return tree;
+}
+
+function flattenSubTree(t, prefix = [], recurse = false) {
+  const directChildren = [...Object.keys(t)]
+    .filter(
+      (k) =>
+        ![HTML_FILENAME, MARKDOWN_FILENAME].includes(k) &&
+        (t[k][HTML_FILENAME] === null || t[k][MARKDOWN_FILENAME] === null)
+    )
+    .map((k) =>
+      [
+        ...prefix,
+        k,
+        t[k][MARKDOWN_FILENAME] === null ? MARKDOWN_FILENAME : HTML_FILENAME,
+      ].join("/")
+    );
+  if (!recurse) {
+    return directChildren;
+  }
+
+  const recursiveChildren = [...Object.entries(t)]
+    .filter(([, v]) => v !== null)
+    .flatMap(([k, v]) => flattenSubTree(v, [...prefix, k], recurse));
+  return [...directChildren, ...recursiveChildren];
+}
+
+function childrenForPath(
+  tree,
+  [current, ...path],
+  recurse = false,
+  prefix = []
+) {
+  if (!current) {
+    return flattenSubTree(tree, prefix, recurse);
+  }
+  const parent = tree[current];
+  if (!parent) {
+    return [];
+  }
+  return childrenForPath(parent, path, recurse, [...prefix, current]);
+}
+
+// When running a production build we can afford to look up all files upfront.
+const ALL_DOCUMENT_PATHS_TREE =
+  process.env.NODE_ENV !== "production"
+    ? {}
+    : {
+        [CONTENT_ROOT]: allDocumentPathsAsTree(CONTENT_ROOT),
+        [CONTENT_TRANSLATED_ROOT]: CONTENT_TRANSLATED_ROOT
+          ? allDocumentPathsAsTree(CONTENT_TRANSLATED_ROOT)
+          : {},
+      };
```

Comment:
Might be more readable to wrap this in an [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) or a separate function `getInitialDocumentPathTree()`?